File: plugins/org.eclipse.tm.tcf.core/src/org/eclipse/tm/internal/tcf/services/remote/LocatorProxy.java
/*******************************************************************************
 * Copyright (c) 2007, 2008 Wind River Systems, Inc. and others.
 * All rights reserved. This program and the accompanying materials 
 * are made available under the terms of the Eclipse Public License v1.0 
 * which accompanies this distribution, and is available at 
 * http://www.eclipse.org/legal/epl-v10.html 
 *  
 * Contributors:
 *     Wind River Systems - initial API and implementation
 *******************************************************************************/
package org.eclipse.tm.internal.tcf.services.remote;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import org.eclipse.tm.tcf.core.Command;
import org.eclipse.tm.tcf.protocol.IChannel;
import org.eclipse.tm.tcf.protocol.IPeer;
import org.eclipse.tm.tcf.protocol.IToken;
import org.eclipse.tm.tcf.protocol.JSON;
import org.eclipse.tm.tcf.protocol.Protocol;
import org.eclipse.tm.tcf.services.ILocator;


public class LocatorProxy implements ILocator {
    
    private final IChannel channel;
    private final Map<String,IPeer> peers = new HashMap<String,IPeer>();
    private final Collection<LocatorListener> listeners = new ArrayList<LocatorListener>();
    
    private boolean get_peers_done = false;
    
    private class Peer implements IPeer {
        
        private final IPeer parent;
        
        private final Map<String, String> attrs;
        
        Peer(IPeer parent, Map<String,String> attrs) {
            this.parent = parent;
            this.attrs = attrs;
        }

        public Map<String, String> getAttributes() {
            assert Protocol.isDispatchThread();
            return attrs;
        }

        public String getID() {
            assert Protocol.isDispatchThread();
            return attrs.get(ATTR_ID);
        }

        public String getName() {
            assert Protocol.isDispatchThread();
            return attrs.get(ATTR_NAME);
        }

        public String getOSName() {
            assert Protocol.isDispatchThread();
            return attrs.get(ATTR_OS_NAME);
        }

        public String getTransportName() {
            assert Protocol.isDispatchThread();
            return attrs.get(ATTR_TRANSPORT_NAME);
        }

        public IChannel openChannel() {
            assert Protocol.isDispatchThread();
            IChannel c = parent.openChannel();
            c.redirect(getID());
            return c;
        }
    };

    private final IChannel.IEventListener event_listener = new IChannel.IEventListener() {

        @SuppressWarnings("unchecked")
        public void event(String name, byte[] data) {
            try {
                Object[] args = JSON.parseSequence(data);
                if (name.equals("peerAdded")) {
                    assert args.length == 1;
                    IPeer peer = new Peer(channel.getRemotePeer(), (Map<String,String>)args[0]);
                    if (peers.get(peer.getID()) != null) {
                        Protocol.log("Invalid peerAdded event", new Error());
                        return;
                    }
                    peers.put(peer.getID(), peer);
                    for (LocatorListener l : listeners.toArray(new LocatorListener[listeners.size()])) {
                        try {
                            l.peerAdded(peer);
                        }
                        catch (Throwable x) {
                            Protocol.log("Unhandled exception in Locator listener", x);
                        }
                    }
                }
                else if (name.equals("peerChanged")) {
                    assert args.length == 1;
                    Map<String,String> m = (Map<String,String>)args[0];
                    if (m == null) throw new Error("Locator service: invalid peerChanged event - no peer ID");
                    IPeer peer = peers.get(m.get(IPeer.ATTR_ID));
                    if (peer == null) return;
                    peers.put(peer.getID(), peer);
                    for (LocatorListener l : listeners.toArray(new LocatorListener[listeners.size()])) {
                        try {
                            l.peerChanged(peer);
                        }
                        catch (Throwable x) {
                            Protocol.log("Unhandled exception in Locator listener", x);
                        }
                    }
                }
                else if (name.equals("peerRemoved")) {
                    assert args.length == 1;
                    String id = (String)args[0];
                    IPeer peer = peers.remove(id);
                    if (peer == null) return;
                    for (LocatorListener l : listeners.toArray(new LocatorListener[listeners.size()])) {
                        try {
                            l.peerRemoved(id);
                        }
                        catch (Throwable x) {
                            Protocol.log("Unhandled exception in Locator listener", x);
                        }
                    }
                }
                else if (name.equals("peerHeartBeat")) {
                    assert args.length == 1;
                    String id = (String)args[0];
                    IPeer peer = peers.get(id);
                    if (peer == null) return;
                    for (LocatorListener l : listeners.toArray(new LocatorListener[listeners.size()])) {
                        try {
                            l.peerHeartBeat(id);
                        }
                        catch (Throwable x) {
                            Protocol.log("Unhandled exception in Locator listener", x);
                        }
                    }
                }
                else {
                    throw new IOException("Locator service: unknown event: " + name);
                }
            }
            catch (Throwable x) {
                channel.terminate(x);
            }
        }
    };
    
    public LocatorProxy(IChannel channel) {
        this.channel = channel;
        channel.addEventListener(this, event_listener);
    }

    public String getName() {
        return NAME;
    }

    public Map<String,IPeer> getPeers() {
        return peers;
    }

    public IToken redirect(String peer_id, final DoneRedirect done) {
        return new Command(channel, this, "redirect", new Object[]{ peer_id }) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error == null) {
                    assert args.length == 1;
                    error = toError(args[0]);
                }
                done.doneRedirect(token, error);
            }
        }.token;
    }

    public IToken sync(final DoneSync done) {
        return new Command(channel, this, "sync", null) {
            @Override
            public void done(Exception error, Object[] args) {
                if (error != null) channel.terminate(error);
                done.doneSync(token);
            }
        }.token;
    }

    public void addListener(LocatorListener listener) {
        listeners.add(listener);
        if (!get_peers_done) {
            new Command(channel, this, "getPeers", null) {
                @SuppressWarnings("unchecked")
                @Override
                public void done(Exception error, Object[] args) {
                    if (error == null) {
                        assert args.length == 2;
                        error = toError(args[0]);
                    }
                    if (error != null) {
                        Protocol.log("Locator error", error);
                        return;
                    }
                    Collection<?> c = (Collection<?>)args[1];
                    if (c != null) {
                        for (Object o : c) {
                            Map<String,String> m = (Map<String,String>)o;
                            String id = m.get(IPeer.ATTR_ID);
                            if (peers.get(id) != null) continue;
                            IPeer peer = new Peer(channel.getRemotePeer(), m);
                            peers.put(id, peer);
                            for (LocatorListener l : listeners.toArray(new LocatorListener[listeners.size()])) {
                                try {
                                    l.peerAdded(peer);
                                }
                                catch (Throwable x) {
                                    Protocol.log("Unhandled exception in Locator listener", x);
                                }
                            }
                        }
                    }
                }
            };
            get_peers_done = true;
        }
    }

    public void removeListener(LocatorListener listener) {
        listeners.remove(listener);
    }
}

File: python/src/tcf/services/remote/LocatorProxy.py
# *******************************************************************************
# * Copyright (c) 2011 Wind River Systems, Inc. and others.
# * All rights reserved. This program and the accompanying materials
# * are made available under the terms of the Eclipse Public License v1.0
# * which accompanies this distribution, and is available at
# * http://www.eclipse.org/legal/epl-v10.html
# *
# * Contributors:
# *     Wind River Systems - initial API and implementation
# *******************************************************************************

import exceptions
from tcf import protocol, peer, channel
from tcf.services import locator
from tcf.channel.Command import Command

class Peer(peer.TransientPeer):
    def __init__(self, parent, attrs):
        super(Peer, self).__init__(attrs)
        self.parent = parent
    def openChannel(self):
        assert protocol.isDispatchThread()
        c = self.parent.openChannel()
        c.redirect(self.getID())
        return c

class ChannelEventListener(channel.EventListener):
    def __init__(self, proxy):
        self.proxy = proxy
        self.channel = proxy.channel
    def event(self, name, data):
        try:
            args = channel.fromJSONSequence(data)
            if name == "peerAdded":
                assert len(args) == 1
                peer = Peer(self.channel.getRemotePeer(), args[0])
                if self.proxy.peers.get(peer.getID()):
                    protocol.log("Invalid peerAdded event", exceptions.Exception())
                    return
                self.proxy.peers[peer.getID()] = peer
                for l in self.proxy.listeners:
                    try:
                        l.peerAdded(peer)
                    except exceptions.Exception as x:
                        protocol.log("Unhandled exception in Locator listener", x)
            elif name == "peerChanged":
                assert len(args) == 1
                m = args[0]
                if not m: raise exceptions.Exception("Locator service: invalid peerChanged event - no peer ID")
                peer = self.proxy.peers.get(m.get(peer.ATTR_ID))
                if not peer: return
                self.proxy.peers[peer.getID()] = peer
                for l in self.proxy.listeners:
                    try:
                        l.peerChanged(peer)
                    except exceptions.Exception as x:
                        protocol.log("Unhandled exception in Locator listener", x)
            elif name == "peerRemoved":
                assert len(args) == 1
                id = args[0]
                peer = self.proxy.peers.get(id)
                if not peer: return
                del self.proxy.peers[id]
                for l in self.proxy.listeners:
                    try:
                        l.peerRemoved(id)
                    except exceptions.Exception as x:
                        protocol.log("Unhandled exception in Locator listener", x)
            elif name == "peerHeartBeat":
                assert len(args) == 1
                id = args[0]
                peer = self.proxy.peers.get(id)
                if not peer: return
                for l in self.proxy.listeners:
                    try:
                        l.peerHeartBeat(id)
                    except exceptions.Exception as x:
                        protocol.log("Unhandled exception in Locator listener", x)
            else:
                raise exceptions.IOError("Locator service: unknown event: " + name)
        except exceptions.Exception as x:
            self.channel.terminate(x)

class LocatorProxy(locator.LocatorService):
    def __init__(self, channel):
        self.channel = channel;
        self.peers = {}
        self.listeners = []
        self.get_peers_done = False
        self.event_listener = ChannelEventListener(self)
        channel.addEventListener(self, self.event_listener)

    def getPeers(self):
        return self.peers

    def redirect(self, peer, done):
        done = self._makeCallback(done)
        service = self
        class RedirectCommand(Command):
            def __init__(self):
                super(RedirectCommand, self).__init__(service.channel, service, "redirect", [peer])
            def done(self, error, args):
                if not error:
                    assert len(args) == 1
                    error = self.toError(args[0])
                done.doneRedirect(self.token, error)
        return RedirectCommand().token

    def sync(self, done):
        done = self._makeCallback(done)
        service = self
        class SyncCommand(Command):
            def __init__(self):
                super(SyncCommand, self).__init__(service.channel, service, "sync", None)
            def done(self, error, args):
                if error: service.channel.terminate(error)
                done.doneSync(self.token)
        return SyncCommand().token

    def addListener(self, listener):
        self.listeners.add(listener)
        if not self.get_peers_done:
            service = self
            class GetPeersCommand(Command):
                def __init__(self):
                    super(GetPeersCommand, self).__init__(service.channel, service, "getPeers", None)
                def done(self, error, args):
                    if not error:
                        assert len(args) == 2
                        error = self.toError(args[0])
                    if error:
                        protocol.log("Locator error", error)
                        return
                    c = args[1]
                    if c:
                        for m in c:
                            id = m.get(peer.ATTR_ID)
                            if service.peers.get(id): continue;
                            peer = Peer(service.channel.getRemotePeer(), m)
                            service.peers[id] = peer
                            for l in service.listeners:
                                try:
                                    l.peerAdded(peer)
                                except exceptions.Exception as x:
                                    protocol.log("Unhandled exception in Locator listener", x)
            self.get_peers_done = True

    def removeListener(self, listener):
        self.listeners.remove(listener)
